src/io/flutter/run/MainFile.java

```diff
@@ -0,0 +1,173 @@
+ * <p>
+ * It has been checked for errors, but the check might not have succeeded.
+ */
+public class MainFile {
+
+  @Nullable
+  private final VirtualFile file;
+
+  @Nullable
+  private final VirtualFile appDir;
+
+  @Nullable
+  private final String error;
+
+  private MainFile(@Nullable VirtualFile file, @Nullable VirtualFile appDir, @Nullable String error) {
+    assert ((file != null && appDir != null) || error != null);
+    this.file = file;
+    this.appDir = appDir;
+    this.error = error;
+  }
+
+  /**
+   * Returns true if the Flutter app can be launched.
+   * <p>
+   * If false, the error can be found by calling {@link #getError}.
+   */
+  public boolean canLaunch() {
+    return error == null;
+  }
+
+  /**
+   * Returns the error message to display if this file is not launchable.
+   */
+  @NotNull
+  public String getError() {
+    if (error == null) {
+      throw new IllegalStateException("called getError when there is no error");
+    }
+    return error;
+  }
+
+  /**
+   * Returns the Dart file containing main.
+   */
+  @NotNull
+  public VirtualFile getFile() {
+    if (file == null) {
+      throw new IllegalStateException("called getFile without checking for error: " + error);
+    }
+    return file;
+  }
+
+  /**
+   * Returns the closest ancestor directory containing a pubspec.yaml, BUILD, or .packages file.
+   */
+  @NotNull
+  public VirtualFile getAppDir() {
+    if (appDir == null) {
+      throw new IllegalStateException("called getAppDir witrhout checking for error: " + error);
+    }
+    return appDir;
+  }
+
+  /**
+   * Verifies that the given path points to an entrypoint file within a Flutter app.
+   * <p>
+   * If there is an error, {@link #canLaunch} will return false and the error is available via {@link #getError}
+   */
+  public static MainFile verify(@Nullable String path, Project project) {
+    if (!ApplicationManager.getApplication().isReadAccessAllowed()) {
+      throw new IllegalStateException("need read access");
+    }
+
+    if (StringUtil.isEmptyOrSpaces(path)) {
+      return error(FlutterBundle.message("entrypoint.not.set"));
+    }
+
+    final VirtualFile file = LocalFileSystem.getInstance().findFileByPath(path);
+    if (file == null) {
+      return error(FlutterBundle.message("entrypoint.not.found", FileUtil.toSystemDependentName(path)));
+    }
+    if (file.getFileType() != DartFileType.INSTANCE) {
+      return error(FlutterBundle.message("entrypoint.not.dart"));
+    }
+
+    final PsiFile psi = PsiManager.getInstance(project).findFile(file);
+    if (!(psi instanceof DartFile)) {
+      return error(FlutterBundle.message("entrypoint.not.dart"));
+    }
+    final DartFile dart = (DartFile)psi;
+
+    if (DartResolveUtil.getMainFunction(dart) == null) {
+      return error(FlutterBundle.message("main.not.in.entrypoint"));
+    }
+
+    if (findImportUrls(dart).noneMatch((url) -> url.startsWith("package:flutter/"))) {
```

Comment:
The code I borrowed from is for running the app by right-clicking, so I think it needs to avoid overriding the regular Dart run config if you're running a non-Flutter main method. I guess it makes sense there but not here.